library/src/main/java/com/google/android/exoplayer/MediaCodecUtil.java

```diff
@@ -95,34 +105,72 @@ public static synchronized void warmCodec(String mimeType, boolean secure) {
-    // TODO: Verify this cannot occur on v22, and change >= to == [Internal: b/18678462].
-    if (secure && codecInfo == null && Util.SDK_INT >= 21) {
-      // Some devices don't list secure decoders on API level 21. Try the legacy path.
-      mediaCodecList = new MediaCodecListCompatV16();
+      String originalMimeType, boolean secure) throws DecoderQueryException {
+
+    // Build a collection of possible mimetypes which we can use to try and identify the associated
+    // MediaCodec.
+    ArrayList<String> mimeTypesToConsider = new ArrayList<>();
+    mimeTypesToConsider.add(originalMimeType);
+    if (mimeTypeAlternatives.containsKey(originalMimeType))
+      mimeTypesToConsider.addAll(mimeTypeAlternatives.get(originalMimeType));
+
+    Pair<String, CodecCapabilities> codecInfo = null;
+    for (String mimeType : mimeTypesToConsider) {
+      CodecKey key = new CodecKey(mimeType, secure);
+      if (codecs.containsKey(key)) {
+        return codecs.get(key);
+      }
+      MediaCodecListCompat mediaCodecList = Util.SDK_INT >= 21
+              ? new MediaCodecListCompatV21(secure) : new MediaCodecListCompatV16();
       codecInfo = getMediaCodecInfo(key, mediaCodecList);
-      if (codecInfo != null) {
-        Log.w(TAG, "MediaCodecList API didn't list secure decoder for: " + mimeType
-            + ". Assuming: " + codecInfo.first);
+      // TODO: Verify this cannot occur on v22, and change >= to == [Internal: b/18678462].
+      if (secure && codecInfo == null && Util.SDK_INT >= 21) {
+        // Some devices don't list secure decoders on API level 21. Try the legacy path.
+        mediaCodecList = new MediaCodecListCompatV16();
+        codecInfo = getMediaCodecInfo(key, mediaCodecList);
+        if (codecInfo != null) {
+          Log.w(TAG, "MediaCodecList API didn't list secure decoder for: " + mimeType
+                  + ". Assuming: " + codecInfo.first);
+        }
+      }
+
+      if (codecInfo != null && mimeTypeAlternatives.containsKey(originalMimeType)) {
+        // The original mime type had a number of possible alternatives. Now that we've determined
+        // the "correct" one, we remove any of the others so that we don't attempt to use them again
+        // in the future.
+        ArrayList<String> alternatives = mimeTypeAlternatives.get(originalMimeType);
+        alternatives.clear();
+        alternatives.add(mimeType);
       }
     }
+
     return codecInfo;
   }
 
+  /**
+   * Returns the actual mime type known by this device for the one provided. This allows the library
+   *     to use a potentially different mime type internally, but translate back to the device
+   *     specific one when required.
+   *
+   * @param originalMimeType The mime type.
+   * @return The devices specific mime type for the one specified.
+   */
+  public static synchronized String getDeviceMimeType(String originalMimeType) {
+    ArrayList<String> alternatives = mimeTypeAlternatives.get(originalMimeType);
```

Comment:
Meh. It's quite a lot more complicated than I thought it was going to be. CodecKey makes things more complicated still, if we want to handle everything in a nice way.

I think the Sony TVs use the same MimeTypes as are currently in place in ExoPlayer these days, although @andrewlewis can confirm/deny. If that's the case, it might be worth holding the thought for a while to see if other mainstream devices do actually ship with the need for this kind of hack. If they do, maybe we should think again at that point?